### PR TITLE
[MSHARED-1330] Always overwrite files

### PR DESCRIPTION
**Pull request by gnodet**
Make sure the files are made writable before copying them.  The attributes are always restored after the copy.

### CODE REVIEW

**Review by elharo — COMMENTED** [review]
why do you need to set the execute bit?

**Comment by gnodet** [comment]
> why do you need to set the execute bit?

We don't, we just need it to be readable and writable, I'll remove that line.

**Comment by gnodet** [comment]
> why do you need to set the execute bit?

Fixed

**Review by cstamas — APPROVED** [review]

File: src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java [hunk 1]
```diff
@@ -371,6 +375,23 @@ private static void copyFilePermissions(File source, File destination) throws IO
         }
     }
 
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    private static void setAllPermissions(File file) throws IOException {
```
Comment (elharo) [inline src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java:379]: makeWriteable as it no longer sets all permissions

File: src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java [hunk 2]
```diff
@@ -371,6 +375,23 @@ private static void copyFilePermissions(File source, File destination) throws IO
...
+        if (file.exists()) {
+            try {
+                Files.setPosixFilePermissions(
+                        file.toPath(),
+                        EnumSet.of(
+                                PosixFilePermission.OWNER_READ, PosixFilePermission.OWNER_WRITE,
+                                PosixFilePermission.GROUP_READ, PosixFilePermission.GROUP_WRITE,
+                                PosixFilePermission.OTHERS_READ, PosixFilePermission.OTHERS_WRITE));
```
Comment (elharo) [inline src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java:387]: I think the copied file is now world readable and writable when everything completes. That might leave arbitrary world writable files sitting around in the file system, which feels risky. Can this work without world or even group permissions? And/or can the file be set back to owner only after the copy finishes?

**Review by elharo — CHANGES_REQUESTED** [review]

File: src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java (hunk 2 above)
Comment (gnodet) [reply src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java:387]: It's not.  The permissions from the source file are copied, see https://github.com/apache/maven-filtering/blob/ac09a5bc42e7fe812b93106537372441a8cbd6b4/src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java#L333
That's why I did not bother initially with the exec flag.

Comment (elharo) [reply src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java:387]: copyFilePermissions might belong in the public copyFile method above then. As is, that's not necessarily called.

Comment (gnodet) [reply src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java:387]: > copyFilePermissions might belong in the public copyFile method above then. As is, that's not necessarily called.

Not sure I understand your objection.  Are you implying that `private static` static methods should always be inlined ?

Comment (elharo) [reply src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java:387]: nvm, the relevant part was hidden by Github's code review folding

File: src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java [hunk 3]
```diff
@@ -371,6 +375,23 @@ private static void copyFilePermissions(File source, File destination) throws IO
         }
     }
 
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    private static void setRwPermissions(File file) throws IOException {
```
Comment (elharo) [inline src/main/java/org/apache/maven/shared/filtering/FilteringUtils.java:379]: setReadWritePermissions (I initially read this as setRawPermissions)

**Review by elharo — APPROVED** [review]

**Review by slawekjaranowski — APPROVED** [review]

**Review by elharo — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #260

**Pull request merged**